### PR TITLE
Expose the metadata key on client interfaces

### PR DESCRIPTION
**Pull request by abrooksv**
## Description
Add a new static field on the client interfaces with metadata ID (aka API endpoint prefix)

## Motivation and Context
#2301 

## Testing
```
public interface ApiGatewayClient extends SdkClient {
    String SERVICE_NAME = "apigateway";

    /**
     * Value for looking up the service's metadata from the
     * {@link software.amazon.awssdk.regions.ServiceMetadataProvider}.
     */
    String SERVICE_METADATA_ID = "apigateway";
```

```
public interface ApiGatewayAsyncClient extends SdkClient {
    String SERVICE_NAME = "apigateway";

    /**
     * Value for looking up the service's metadata from the
     * {@link software.amazon.awssdk.regions.ServiceMetadataProvider}.
     */
    String SERVICE_METADATA_ID = "apigateway";
```

```
public interface EcrAsyncClient extends SdkClient {
    String SERVICE_NAME = "ecr";

    /**
     * Value for looking up the service's metadata from the
     * {@link software.amazon.awssdk.regions.ServiceMetadataProvider}.
     */
    String SERVICE_METADATA_ID = "api.ecr";
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json/service-2.json
Comment (abrooksv) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json/service-2.json:11]: Added so it tests when singing name != endpoint prefix

**Comment by abrooksv** [comment]
I am a bit stumped why the sync client code gen test is failing with extra spaces in the JavaDoc, ideas

Screen Shot 2021-03-11 at 1 23 54 PM

**Comment by abrooksv** [comment]
Digging into this more it seems there is a 

```
    static ServiceMetadata serviceMetadata() {
        return ServiceMetadata.of(SERVICE_METADATA_ID);
    }
```
on the sync interface (it is missing from the async interface).

This is basically what I needed the ID for in the first place, but on order to make what I need work (dynamic service info in the IDEs), I basically need to find a way to update `MetadataLoader` with an updated endpoints.json data that the toolkits team vends from CloudFront (https://idetoolkits.amazonwebservices.com/endpoints.json)

```
/**
 * Internal class for determing where to load region and service
 * metadata from. Currently only generated region metadata is supported.
 */
@SdkInternalApi
final class MetadataLoader {
```

Is there any thoughts on how to make the `Currently only generated region metadata is supported.` a bit less true of a statement 😉 To be clear, I am not advocating for dropping the codegen, but a way to load newer data into the `MetadataLoader` using the SDKs endpoints metadata format. Note: The toolkits are onboarded to the SDK release platform, that is how we get this data today. 

The endpoints.json parser currently lives in `codegen-lite`, so it would have to be added to new module that lives below `codegen-lite` and `regions`.

**Comment by Quanzzzz** [comment]
Sorry for delay. This pr should be good to go once we get another review from the team, as this is a change to the public apis. About the extra feature request mentioned in the comment, we can implement a new utillity method to read the service metadata directly from the json file so you can get "dynamic service info". We've created a new feature request task in our backlog, will implement it once it's prioritized.

**Review by Quanzzzz — APPROVED** [review]

**Comment by zoewangg** [comment]
@all-contributors please add @abrooksv for code

**Pull request merged**
